<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MRLocus - Mendelian Randomization per locus • mrlocus</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="MRLocus - Mendelian Randomization per locus">
<meta property="og:description" content="Mendelian Randomization per locus, leveraging eQTL
    and GWAS summary statistics, for estimation of gene-to-trait
    effect size and dispersion.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">mrlocus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.22</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/mrlocus.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mikelove/mrlocus/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="mrlocus" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#mrlocus" class="anchor"></a>MRLocus</h1></div>

<div id="about" class="section level3">
<h3 class="hasAnchor">
<a href="#about" class="anchor"></a>About</h3>
<p>MRLocus is a statistical method providing Mendelian Randomization analysis per locus, leveraging eQTL and GWAS summary statistics, for estimation of gene-to-trait effect size and dispersion.</p>
<p>See <code>Get started</code> tab for software vignette, and <code>Reference</code> tab for manual pages of the individual functions.</p>
</div>
<div id="preprint" class="section level3">
<h3 class="hasAnchor">
<a href="#preprint" class="anchor"></a>Preprint</h3>
<p>Anqi Zhu*, Nana Matoba*, Emmaleigh Wilson, Amanda L. Tapia, Yun Li, Joseph G. Ibrahim, Jason L. Stein, Michael I. Love. “MRLocus: identifying causal genes mediating a trait through Bayesian estimation of allelic heterogeneity” <em>bioRxiv</em>, August 2020. <a href="https://doi.org/10.1101/2020.08.14.250720">doi: 10.1101/2020.08.14.250720</a></p>
<p>* Both authors contributed equally to this work.</p>
</div>
<div id="data-input" class="section level3">
<h3 class="hasAnchor">
<a href="#data-input" class="anchor"></a>Data input</h3>
<p>MRLocus uses multiple, distinct eQTL signals in order to perform Mendelian Randomization investigating the effect of gene expression on a downstream GWAS trait. It takes as input the output of a clumping procedure, such as <a href="https://zzz.bwh.harvard.edu/plink/clump.shtml">PLINK clump</a>, or ideally, <a href="https://pubmed.ncbi.nlm.nih.gov/28165122/">conditional analysis</a> of the eQTL dataset. In the MRLocus paper, we use the PLINK clump settings, <code>--clump-p1 0.001 --clump-p2 1 --clump-r2 0.1 --clump-kb 500</code>, where <code>p1</code> refers to the eQTL summary associations. MRLocus can then be run on loci that contain more than one clump.</p>
<p>MRLocus then requires:</p>
<ol>
<li>Summary statistics for each clump</li>
<li>Intra-clump LD matrices</li>
</ol>
<p>Summary statistics for both eQTL and GWAS (effect size, standard error, reference allele, effect allele) of the SNPs in each clump or conditionally distinct cluster should be gathered into tables, making sure the index eSNP is also included. The eQTL and GWAS studies can have different coding of reference and effect alleles, MRLocus has a function that will deal with this internally, to avoid users having to flip alleles and effect sizes themselves.</p>
<p>After having prepared these multiple tables for each clump, they should be read into R as a list of <code>data.frame</code>. The LD matrices for each clump likewise should be read in as a list of matrices (in the same order as the summary statistics list). These lists can be read in with code such as:</p>
<pre><code>sum_stat &lt;- lapply(sum_stat_files, read.delim)
ld_mat &lt;- lapply(ld_mat_files, function(f) as.matrix(read.table(f)))</code></pre>
<p>The steps following reading in the summary statistics and LD matrices into R are described in the vignette under the <code>Get started</code> tab above.</p>
<p>Finally, we note that MRLocus offers its own colocalization model, or alternatively, MRLocus slope fitting can be run after performing <a href="https://pubmed.ncbi.nlm.nih.gov/27866706/">eCAVIAR</a> colocalization on the SNPs within each clump. The alternative steps used in the case of eCAVIAR colocalization are discussed in the vignette under <code>Get started</code> tab.</p>
</div>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<p>The <em>mrlocus</em> R package can be installed using <em>devtools</em>. Note that <em>mrlocus</em> uses RStan, and so involves compilation of C++ code. Installation will take longer than with R packages not containing code to be compiled.</p>
<pre><code><a href="https://devtools.r-lib.org/">library(devtools)
install_github("mikelove/mrlocus")</a></code></pre>
</div>
<div id="issues" class="section level3">
<h3 class="hasAnchor">
<a href="#issues" class="anchor"></a>Issues</h3>
<p>For bug reports, feature requests, or questions, please post <a href="https://github.com/mikelove/mrlocus/issues/new/choose">here</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/mikelove/mrlocus/">https://​github.com/​mikelove/​mrlocus/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;=2)</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing mrlocus</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Anqi Zhu <br><small class="roles"> Author </small>  </li>
<li>Nana Matoba <br><small class="roles"> Author </small>  </li>
<li>Jason Stein <br><small class="roles"> Author </small>  </li>
<li>Michael Love <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.com/mikelove/mrlocus"><img src="https://travis-ci.com/mikelove/mrlocus.svg?branch=master" alt="Build Status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Anqi Zhu, Nana Matoba, Jason Stein, Michael Love.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
