<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Extract SNPs from colocalization for slope fitting — extractForSlope • mrlocus</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Extract SNPs from colocalization for slope fitting — extractForSlope"><meta property="og:description" content="Extracts one or more SNPs from each signal cluster
based on the posterior estimate of the effect size
for A (largest effect size in the positive direction).
After running this function, it is recommended to use
trimClusters to remove signal clusters
that are too highly correlated."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mrlocus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/mrlocus.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mikelove/mrlocus/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract SNPs from colocalization for slope fitting</h1>
    <small class="dont-index">Source: <a href="https://github.com/mikelove/mrlocus/blob/HEAD/R/helper.R" class="external-link"><code>R/helper.R</code></a></small>
    <div class="hidden name"><code>extractForSlope.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Extracts one or more SNPs from each signal cluster
based on the posterior estimate of the effect size
for A (largest effect size in the positive direction).
After running this function, it is recommended to use
<code><a href="trimClusters.html">trimClusters</a></code> to remove signal clusters
that are too highly correlated.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extractForSlope</span><span class="op">(</span></span>
<span>  <span class="va">res</span>,</span>
<span>  niter <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label <span class="op">=</span> <span class="st">"Effect size of"</span>,</span>
<span>  a <span class="op">=</span> <span class="st">"eQTL"</span>,</span>
<span>  b <span class="op">=</span> <span class="st">"GWAS"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>res</dt>
<dd><p>list with the following named elements:</p><ul><li><p><code>beta_hat_a</code> - list of point estimates of coefficients for A from colocalization</p></li>
<li><p><code>beta_hat_b</code> - " " for B</p></li>
<li><p><code>sd_a</code> - list of sampling SD for <code>beta_hat_a</code> (in practice original
SE are provided here)</p></li>
<li><p><code>sd_b</code> - " " for <code>beta_hat_b</code> " "</p></li>
<li><p>alleles (optional) list of data.frame with allele information</p></li>
</ul></dd>


<dt>niter</dt>
<dd><p>number of iterations of EM to run
for mclust, if set to 0, only the maximum
variant (in terms of A effect size) per
signal cluster is output. Default is to not
run clustering, but to take the SNP with the
largest effect size in A (in the positive direction)</p></dd>


<dt>plot</dt>
<dd><p>logical, draw a before after of which
variants will be included for slope estimation</p></dd>


<dt>label</dt>
<dd><p>what preceeds <code>a</code> and <code>b</code> in
the x- and y-axis labels</p></dd>


<dt>a</dt>
<dd><p>name of A experiment</p></dd>


<dt>b</dt>
<dd><p>name of B experiment</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>list of vectors of the first four arguments,
collapsed now across signal clusters, representing
variants with positive effect on A. So the null variants
have been removed (and any variants per cluster that
indicated a negative effect on A). If <code>alleles</code></p>


<p>data.frames were included in the input, they will
also be passed through as a single data.frame with the
selected SNPs per signal cluster</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anqi Zhu, Nana Matoba, Jason Stein, Michael Love.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

