<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fit the colocalization model per signal cluster — fitBetaColoc • mrlocus</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fit the colocalization model per signal cluster — fitBetaColoc" />
<meta property="og:description" content="Implements the MRLocus colocalization step, in which summary
statistics from A and B studies are used in a generative
model, with a horseshoe prior on the latent true effect sizes.
Posterior effect sizes and posterior SD are returned,
although in pratice we use the original SE in the subsequent
fitSlope model.
See Supplementary Methods of the MRLocus manuscript." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-175865595-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-175865595-1');
</script>


  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mrlocus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/mrlocus.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/mikelove/mrlocus/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit the colocalization model per signal cluster</h1>
    <small class="dont-index">Source: <a href='https://github.com/mikelove/mrlocus/blob/master/R/mrlocus.R'><code>R/mrlocus.R</code></a></small>
    <div class="hidden name"><code>fitBetaColoc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Implements the MRLocus colocalization step, in which summary
statistics from A and B studies are used in a generative
model, with a horseshoe prior on the latent true effect sizes.
Posterior effect sizes and posterior SD are returned,
although in pratice we use the original SE in the subsequent
<code><a href='fitSlope.html'>fitSlope</a></code> model.
See Supplementary Methods of the MRLocus manuscript.</p>
    </div>

    <pre class="usage"><span class='fu'>fitBetaColoc</span>(<span class='no'>beta_hat_a</span>, <span class='no'>beta_hat_b</span>, <span class='no'>se_a</span>, <span class='no'>se_b</span>, <span class='no'>Sigma_a</span>, <span class='no'>Sigma_b</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>beta_hat_a</th>
      <td><p>vector of estimated coefficients for A</p></td>
    </tr>
    <tr>
      <th>beta_hat_b</th>
      <td><p>" " for B</p></td>
    </tr>
    <tr>
      <th>se_a</th>
      <td><p>vector of standard errors for beta_hat_a</p></td>
    </tr>
    <tr>
      <th>se_b</th>
      <td><p>" " for beta_hat_b</p></td>
    </tr>
    <tr>
      <th>Sigma_a</th>
      <td><p>correlation matrix of SNPs for A</p></td>
    </tr>
    <tr>
      <th>Sigma_b</th>
      <td><p>" " for B (this could be different for different LD matrix)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>further arguments passed to rstan::sampling</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a list with the following elements:</p><ul>
<li><p><code>stanfit</code> object</p></li>
<li><p>posterior means for estimated coefficients for A and B</p></li>
<li><p>posterior standard deviations for A and B</p></li>
<li><p>scaling factors for A and B</p></li>
</ul><p>Two important notes: (1) in MRLocus manuscript, original
SE are used instead of posterior SD in the slope fitting step,
(2) the posterior means and SD for estimated coefficients 
are appropriately scaled, while the results from the
<code>stanfit</code> object are not scaled. In order to scale the
results from the stanfit object, <code>scale_a</code> and <code>scale_b</code>
should be divided out from both coefficients and SDs
(see Supplementary Methods).</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Anqi Zhu*, Nana Matoba*, Emmaleigh Wilson, Amanda L. Tapia, Yun Li,
Joseph G. Ibrahim, Jason L. Stein, Michael I. Love.
MRLocus: identifying causal genes mediating a trait through
Bayesian estimation of allelic heterogeneity. (2020) bioRxiv
<a href='https://doi.org/10.1101/2020.08.14.250720'>https://doi.org/10.1101/2020.08.14.250720</a></p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Anqi Zhu, Nana Matoba, Jason Stein, Michael Love.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


