<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Flip alleles and gather results into lists — flipAllelesAndGather • mrlocus</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Flip alleles and gather results into lists — flipAllelesAndGather"><meta property="og:description" content="A helper function to flip alleles from eQTL and GWAS
datasets, such that they agree on the effect allele,
that the SNPs in a signal cluster are in postive
correlation with the index SNP (eQTL), and that the
effect allele is coded such that it is the expression
increasing allele.
This is recommended to run after
collapseHighCorSNPs, and before
fitBetaColoc."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mrlocus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.25</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/thelovelab/mrlocus/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Flip alleles and gather results into lists</h1>
    <small class="dont-index">Source: <a href="https://github.com/thelovelab/mrlocus/blob/HEAD/R/helper.R" class="external-link"><code>R/helper.R</code></a></small>
    <div class="hidden name"><code>flipAllelesAndGather.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A helper function to flip alleles from eQTL and GWAS
datasets, such that they agree on the effect allele,
that the SNPs in a signal cluster are in postive
correlation with the index SNP (eQTL), and that the
effect allele is coded such that it is the expression
increasing allele.
This is recommended to run after
<code><a href="collapseHighCorSNPs.html">collapseHighCorSNPs</a></code>, and before
<code><a href="fitBetaColoc.html">fitBetaColoc</a></code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">flipAllelesAndGather</span><span class="op">(</span></span>
<span>  <span class="va">sum_stat</span>,</span>
<span>  <span class="va">ld_mat</span>,</span>
<span>  ld_mat2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">a</span>,</span>
<span>  <span class="va">b</span>,</span>
<span>  <span class="va">ref</span>,</span>
<span>  <span class="va">eff</span>,</span>
<span>  <span class="va">beta</span>,</span>
<span>  <span class="va">se</span>,</span>
<span>  <span class="va">a2_plink</span>,</span>
<span>  a2_plink_mat2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">snp_id</span>,</span>
<span>  <span class="va">sep</span>,</span>
<span>  ab_last <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  alleles_same <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sum_stat</dt>
<dd><p>list of summary statistic tables.
A list over signal clusters, where each element is a
data.frame with summary statistics from eQTL and GWAS
datasets. The names of the columns are specified by
arguments below (e.g. <code>a</code>, <code>b</code>, <code>ref</code>,
<code>eff</code>, etc.)</p></dd>


<dt>ld_mat</dt>
<dd><p>list of LD matrices</p></dd>


<dt>ld_mat2</dt>
<dd><p>optional second list of LD matrices
(for different populations). it will be returned
alongside the first <code>ld_mat</code>, which is used
for the allele flipping. the second list of LD
matrices is just flipped in the same way</p></dd>


<dt>a</dt>
<dd><p>name of A in columns of <code>sum_stat</code> ("eQTL")</p></dd>


<dt>b</dt>
<dd><p>name of B ("GWAS")</p></dd>


<dt>ref</dt>
<dd><p>name of reference allele</p></dd>


<dt>eff</dt>
<dd><p>name of effect allele</p></dd>


<dt>beta</dt>
<dd><p>name of estimated coefficient</p></dd>


<dt>se</dt>
<dd><p>name of standard error</p></dd>


<dt>a2_plink</dt>
<dd><p>name of the column representing the a2 allele
(reference allele) according to plink. the default for plink v1.9
and earlier was to reset a2 to the major allele, unless
an optional flag was used. in plink v2.0 and onward, one
should check to see which allele is used as reference for
calculating the LD matrix.
If plink was not used, this argument should just point
to the reference allele that was used for LD calculation</p></dd>


<dt>a2_plink_mat2</dt>
<dd><p>name of the column representing
the a2 allele for the second LD matrix, <code>ld_mat2</code>
(needed only if <code>ld_mat2</code> was specified)</p></dd>


<dt>snp_id</dt>
<dd><p>name of SNP id</p></dd>


<dt>sep</dt>
<dd><p>character separator in column names that involve A/B</p></dd>


<dt>ab_last</dt>
<dd><p>logical, A/B descriptor is last in column names
(e.g. "beta_eqtl", "se_eqtl"))</p></dd>


<dt>alleles_same</dt>
<dd><p>logical, A/B/LD matrix alleles are identical</p></dd>


<dt>plot</dt>
<dd><p>logical, draw a scatterplot of the flipped betas</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>list with estimated coefficients, standard
errors, LD matrix, and alleles data.frame</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Anqi Zhu, Nana Matoba, Jason Stein, Michael Love.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

